### ISSUE TITLE
Using `Breakage` CI for KeywordCalls

### ISSUE THREAD

**Issue by cscherrer** [JuliaAstro/Transits.jl#24]
In case you haven't seen it, this is really helpful:
https://github.com/abelsiqueira/Breakage

It gives you results like this:
https://github.com/cscherrer/KeywordCalls.jl/actions/runs/888461853

For this to be really useful, tests in downstream packages (Transits and MeasureTheory) should be set up to confirm that functionality depending on KeywordCalls is working correctly.

Maybe this is already the case? If that's so, this can be closed right away. The hope is just to avoid having to bug you to check downstream functionality manually. Probably not such a big deal, since KeywordCalls is getting pretty stable now (other than some weirdness with default values). But it's still nice to have one more check on stability :)

**Comment by icweaver** [comment]
Oh that's really neat!

I just realized that I was using our old manual way of constructing objects in our tests instead of with KeywordsCalls. That should be switched over now (just a quick change), is there anything else that we should add on our end for Breakage to pick up on?

**Comment by cscherrer** [comment]
Yeah, Breakage is really nice. As far as anything else to add... I added some tests in KeywordCalls like this:
```julia
@test 0 == @ballocated f(alpha=1,b=2,c=3)
```

Maybe something like that, and one or two with `@belapsed` and some reasonable threshhold? That would make it easy to check that precompilation doesn't slow things down

**Comment by icweaver** [comment]
Great call (hah). I still need to add some extra tests to up our coverage in general, but the additional tests checking the memory allocation and elapsed time using KeywordCalls should be up now and look to be working well locally for me! I will be sure to give an update once the registry catches up with your latest changes

**Comment by icweaver** [comment]
Oh weird, it's not the registry, since `cscherrer/cs-function-resolution` was already merged before the update, IIUC. For some reason, I seem to be seeing allocations in the CI for Julia 1.5+ across platforms, even though KeywordCalls is not locating when I test it locally. Will double check the setup I am using

**Comment by cscherrer** [comment]
Good to know. I had been seeing these some of this in MeasureTheory too. Can you try with this?
https://github.com/cscherrer/KeywordCalls.jl/tree/cs-instance-type

It seemed better behaved to me, but I'm not sure it's yet 100%

**Comment by icweaver** [comment]
hmmm, still seems to have the same issue on the *nix platforms. Will check for Windows next

**Comment by cscherrer** [comment]
I think I need to learn how to use SnoopCompile. You ever tried it?
https://timholy.github.io/SnoopCompile.jl/stable/

**Comment by icweaver** [comment]
Same issue in Windows as well for Julia 1.5+. Was also able to reproduce locally for Julia 1.5, nightly on linux. I guess 1.6 was our sweet spot, although it's a little strange that it only fails remotely

**Comment by icweaver** [comment]
Oh, I haven't. Hopefully Miles has some experience with it!

**Comment by cscherrer** [comment]
Oh 1.6 is ok? I was seeing trouble on 1.4 and 1.5 for a while now

**Comment by icweaver** [comment]
Right, I meant v1.5, 1.6, and nightly CI are failing for me, but 1.6 on my local machine seems to be ok at least (Linux Mint 20.1 x86_64). Strangely enough, v1.0 seems to be passing on all platforms

**Comment by cscherrer** [comment]
Oh good! That's consistent with what I was seeing. That's why I have this line in the tests:
https://github.com/cscherrer/KeywordCalls.jl/blob/cce1b93b913b92c8419eb52aa6b97e2cd44efa2f/test/runtests.jl#L61

**Comment by icweaver** [comment]
Oh wow, I did not realize non-allocating was specific to 1.6, adding that in now! Just out of curiosity, what feature does KeywordCalls use to accomplish this?

**Comment by cscherrer** [comment]
There's not a specific feature. It was just working great in 1.6 but not in earlier versions. I put the conditional in the tests to make sure we keep the 1.6 performance. It would be great is someone finds a way to make this better for 1.4-1.5 without compromising 1.6, but I don't think I'll spend more time on that until there's new information.

**Comment by icweaver** [comment]
I see, definitely reasonable!
